### ISSUE TITLE
Disconnect events don't get triggered on the server when client closes the connection abruptly

### ISSUE THREAD

**Issue by ddehghan** [miguelgrinberg/Flask-SocketIO#666]
Hi Miguel,   

First of all that you for the great work. My goal was to get a list of active users using our service. As a test I added the SID of the user to a global variable on connect and removed the SID on disconnect event.

I implemented this in the example app on this repo and then I presses control R repeatedly and noticed that the array of SIDs keeps growing because then number of disconnects is lower than connect and it keeps growing. 

I was running python 3.6.4 and eventlet.
I switched to gevent and gevent-socket. Still the same issue. 

```

r = redis.StrictRedis(host='redis', port=6379, db=0)

async_mode = "gevent"
socketio = SocketIO(app, async_mode=async_mode)

@socketio.on('connect', namespace='/test')
def test_connect():
    global thread
    r.set("active_users-sid-%s" % request.sid, 1)
    emit('my_response', {'data': 'Connected', 'count': 0})


@socketio.on('disconnect', namespace='/test')
def test_disconnect():
    r.delete("active_users-sid-%s" % request.sid)
    print('Client disconnected', request.sid)


if __name__ == '__main__':
    socketio.run(app, host="0.0.0.0", port=5000, debug=True, use_reloader=False)

```

```
On the client  test both with no options set and with only socketio. Same result. The 

       {
            secure: true,
            upgrade: true,
            rememberUpgrade: true,
            agent: true,
            autoConnect: false,
            transports: ['websocket']
            // transports: ['polling', 'websocket']
        }



Reading the previous bugs it seems there was an issue with disconnect that was fixed in engine.io 1.5. Is my scenario above expected to work? I was just trying to simulate lots of connections that drop out and are unreliable (given that my test only covered the initial connection phase)

Would it help if I sent you a fork of the example app with these changes incorporated?

Thanks again

**Comment by miguelgrinberg** [comment]
Did you want a couple of minutes? Long polling clients are disconnected after a timeout, they can't be handled immediately like WebSocket.

**Comment by ddehghan** [comment]
I did wait many minutes even a day. In fact I used websocket and after refreshing the page multiple times I still get these orphaned SIDs. 

I am pretty sure these connections are dead already.

Is there a way to query the list of open connections on the server?

**Comment by ddehghan** [comment]
ok I found the issue. the problem is here

socketio = SocketIO(app, async_mode="gevent", logger=True)

If I remove the logger=True the connections get closed properly so it has something to do with logging.

Also I made a clean docker container for the example. this will make the process on onboarding new users to this project easier and gives them a working example right a way. It took me so many hours to get to a point that I could get the thing running. I will send you a pull request.

**Comment by jianglijie** [comment]
i get the same problem, i use flask-socketio for server and test by my android phone.
the steps:
1. android app connect to server
2.open Airplane Mode for close network
3.close Airplane Mode
when connect to server after close Airplane Mode,i can see connect with a new sid and disconnect last sid.but the status is ESTABLISHED still in the system.
This problem caused the num of ESTABLISHED increasing. i have to timed restart the server .How to solve this problem

**Comment by paredesivan** [comment]
@noirbizarre please can you help me, can you send me an example of how everything was working?

**Comment by miguelgrinberg** [comment]
I believe this has been address in the current release of python-engineio. Please reopen if the problem persists.

**Comment by nick-landes** [comment]
Hi! Hope you can help me. Having the same issue. When client disconnects just by closing or reloading a tab, server doesn't get triggered on 'disconnect' event. Is it okay and how can I handle this issue?

**Comment by miguelgrinberg** [comment]
@nick-landes Is this a polling or WebSocket client? If polling, it can take up to a minute for the disconnection to be detected.

**Comment by nick-landes** [comment]
@miguelgrinberg Thank you for your reply! If I understand you correctly, I am using socket.io-client. To establish a connection I tried to use: 
const socket = io('url') 
and
const socket = io(url, { transports: [websocket], upgrade: false }) But 'disconnect' doesn't get triggered on server-side even after one minute.

**Comment by miguelgrinberg** [comment]
@nick-landes does your connect handler run? If yes, are you returning anything in this function?

**Comment by nick-landes** [comment]
@miguelgrinberg no I only use 'disconnect' handler to catch client disconnect and socketio.emit() in threads

**Comment by miguelgrinberg** [comment]
@nick-landes Do you think you can provide a small but complete example that demonstrates this? Based on your answers you are not doing anything wrong, but yet I cannot reproduce this problem with my own code.

**Comment by nick-landes** [comment]
@miguelgrinberg 
Here's logic of my app:

import json
from threading import Thread

import cv2
import requests
from flask import Flask
from flask import request
from flask_socketio import SocketIO, disconnect
from flask_cors import CORS

app = Flask(__name__)
CORS(app)
socketio = SocketIO(app, logger=True, engineio_logger=True, cors_allowed_origins='*')


def init_model(data, namespace):
    model_url_predict = f"{data['nnmodel-url']}/predict"

    stream = cv2.VideoCapture(data['source-url'])

    while True:
        model_prediction = requests.post(model_url_predict,
                                         json=json.dumps(stream_data),
                                         headers=headers).json()
        socketio.emit("data", model_prediction,
                      namespace=namespace)


@app.route('/get-source-data', methods=['GET', 'POST'])
def get_source_data():
    global action, status, namespaces
    message = request.get_data()
    message_data = json.loads(message)

    namespace = message_data['source-url'].split('/')[-1]

    thread = Thread(target=init_model, args=(message_data, namespace), daemon=True)
    thread.start()

    status = 'ok'
    return status


@socketio.on('disconnect')
def client_disconnect():
    print('Client disconnected')
    disconnect()


if __name__ == "__main__":
    socketio.run(app, host='0.0.0.0', port=8070, debug=True)

**Comment by nick-landes** [comment]
@miguelgrinberg Here's simplified code off my app. I left logic so you could understand. Thank you.

<img width="608" alt="image" src="https://user-images.githubusercontent.com/22571093/79877627-20476f00-83f5-11ea-8dd6-b6d68bc0e276.png">

**Comment by miguelgrinberg** [comment]
I need actual code I can run please. Can you provide the above server as a file and the client that connects to it?

**Comment by johaven** [comment]
I face the same problem.

Disconnect event is correctly triggered when i close the browser or the browser tab but not triggered if i refresh the page.

i checked **socketio.server.eio.sockets**, the socket is well absent i don't understand why disconnect event was not propagated.

I noticed this because I store the request.sid (append / pop) and for some users I saw 2/3 sids (the disconnect method should have remove the entries).
